package-manifest.toml

```diff
@@ -1,22 +1,36 @@
-[package.crucible]
-service_name = "crucible"
+[package.crucible-svc]
+service_name = "crucible-svc"
 source.type = "local"
 source.rust.binary_names = ["crucible-agent", "crucible-downstairs"]
 source.rust.release = true
 source.paths = [
   { from = "agent/smf", to = "/var/svc/manifest/site/crucible" },
   { from = "agent/downstairs_method_script.sh", to = "/opt/oxide/lib/svc/manifest/crucible/downstairs.sh" },
-  { from = "agent/agent_method_script.sh", to = "/opt/oxide/lib/svc/manifest/crucible/agent.sh" }
 ]
 output.type = "zone"
+output.intermediate_only = true
 
-[package.crucible-pantry]
-service_name = "pantry"
+[package.crucible]
+service_name = "crucible"
+only_for_targets.image = "standard"
+source.type = "composite"
+source.packages = [ "crucible-svc.tar.gz", "zone-network-setup.tar.gz" ]
```

Comment:
Heya @smklein, I have a couple of questions around packages in this repo

1. The zone-network-setup.tar.gz package will be built in omicron https://github.com/oxidecomputer/omicron/pull/4677/files#diff-3ef35f168f90144ed9f4e1c80d4e7b95e4584652bf7f22886046a11c7ef630a6R617-R627 . From my understanding, what I have here will work once deployed on the rack (and the `led-agent/src/services.rs` file is updated of course), but will forever make the rbuild tests fail here? 
  If all of that is wrong, will I have to recreate the `zone-network-setup` service here? :/
2.  I'm not sure what's going on with the packages. I have defined the composite packages here, just like I did in omicron, but for some reason it doesn't create the new composite packages when I run `cargo run --bin crucible-package` :woman_shrugging:

Reply:
> Heya @smklein, I have a couple of questions around packages in this repo
> 
> 1. The zone-network-setup.tar.gz package will be built in omicron https://github.com/oxidecomputer/omicron/pull/4677/files#diff-3ef35f168f90144ed9f4e1c80d4e7b95e4584652bf7f22886046a11c7ef630a6R617-R627 . From my understanding, what I have here will work once deployed on the rack (and the `led-agent/src/services.rs` file is updated of course), but will forever make the rbuild tests fail here?
>    If all of that is wrong, will I have to recreate the `zone-network-setup` service here? :/

I'm not really familiar with what's going on in the rbuild tests in crucible, so I can't really comment on that -- but if this issue is basically a question of dependencies (namely: Omicron generally pulls in Crucible, but here, Crucible wants to pull in something from Omicron, in the form of the network setup), then I think there are two options:

1. Crucible emits zones that are missing their network config. Omicron adds them in during package assembly.

Pro: I *think* this would work with minimal setup
Con: Crucible's zones, without this extra step, would be broken out-of-the-box, because they'd be missing network config (seems liket his might be what you're alluding to with the failing tests?)

2. We could pull the `zone-network-setup` out of Omicron, and have *both* crucible and Omicron pull it in?

I don't think we should duplicate the service, but this would be my standard for breaking out of a circular dependency. We did something similar for the Omicron packaging tools, so they could be used on both sides: https://github.com/oxidecomputer/omicron-package


> 2. I'm not sure what's going on with the packages. I have defined the composite packages here, just like I did in omicron, but for some reason it doesn't create the new composite packages when I run `cargo run --bin crucible-package` 🤷‍♀️

This is the entirety of "crucible-package": https://github.com/oxidecomputer/crucible/blob/e71b10d2f9f1fb52818b916bae83ba15a339548d/package/src/main.rs#L9-L23

What is getting outputted when you run that command?

Reply:
Elaborating on that `create_for_target` call:

- It calls this: https://github.com/oxidecomputer/omicron-package/blob/f581faf67ca74e76d17fa830f27742356be280ef/src/package.rs#L530
- Which would call `create_zone_package` if your output is "zone": https://github.com/oxidecomputer/omicron-package/blob/f581faf67ca74e76d17fa830f27742356be280ef/src/package.rs#L539
- ... and if you have labelled this as a composite package, it shoves all these files into an archive for you: https://github.com/oxidecomputer/omicron-package/blob/f581faf67ca74e76d17fa830f27742356be280ef/src/package.rs#L741-L748

Reply:
Also, just to confirm: The "omicron" repo is using the same version of `omicron-package` as Crucible, so this implementation of building composites should be the same code we're using in Omicron's repo

Reply:
> Pro: I think this would work with minimal setup
Con: Crucible's zones, without this extra step, would be broken out-of-the-box, because they'd be missing network config (seems liket his might be what you're alluding to with the failing tests?)

Yeah, that's what I was alluding to with the failing tests. If we choose to go this way we'd have to make significant changes to the rbuild test here.

> We could pull the zone-network-setup out of Omicron, and have both crucible and Omicron pull it in?

The more I think about it, this is probably the best way to go. It's extremely likely that we'll have more shared services in the future. For example, clickhouse, ch-keeper and cockroachdb all use the [internal-dns](https://github.com/oxidecomputer/omicron/tree/main/smf/internal-dns) service, and a tiny service that disables ssh has just been introduced https://github.com/oxidecomputer/omicron/pull/4716. It's probably best to extract shared services into their own repo it seems :thinking: 

Do you have any thoughts on this @leftwo ?

Reply:
It seems like moving common code to a common location is the way to go, and I don't have any attachment to it being it this repo.

Reply:
Is there any reason that the final assembly cannot be done in omicron, keeping the zone network service and all of its illumos dependencies there?

I haven't tried, but I think this is could work if the service names here were changed slightly (`crucible-zone`, `crucible-pantry-zone`?). Then omicron could continue downloading those with their new names and combine them with the new zone network service to produce the final `crucible` and `crucible-pantry`.

If that's workable, I think it's simpler and keeps final zone assembly in one place.

Reply:
Definitely worth looking into! I'll give it a go